README.md
<div align="center">
  <img src="assets/ContentChef_logo_banner.svg" height="128"/>
</div>

![buildStatusBadge](https://github.com/contentchef/contentchef-jvm/workflows/build/badge.svg)

# ContentChef Java/Kotlin/Android SDK

Welcome to [ContentChef API-First CMS's](https://www.contentchef.io/) Java/Kotlin/Android SDK.

## How to use it

**Java/Kotlin/Android**

Create your ContentChef instance like this:

    val contentChef = CallbackContentChefProvider.getContentChef(  
        ContentChefEnvironmentConfiguration(  
            ContentChefEnvironment.LIVE, SPACE_ID
      )  
    )

*SPACE_ID* can be retrieved from your [ContentChef's dashboard](https://app.contentchef.io/).

You can now use your `contentChef` instance to get the channel you want to use to retrieve info: you have two channels, the `OnlineChannel` and the `PreviewChannel`.

With the `OnlineChannel` you can retrieve contents which are in _live_ state and which are actually visible, while with the `PreviewChannel` you can retrieve contents which are in in both _stage_ and _live_ states and even contents that are not visible in the current date.

Both the `OnlineChannel` and the `PreviewChannel` have two methods which are `getContent()` and `search()`

You can use the `getContent()` method to collect a specific content by its own `publicId`, for example to retrieve a single post from your blog, a single image from a gallery or a set of articles from your featured articles list. Otherwise you can use the `search()` method to find contents with multiple matching criteria, like content definition name, publishing dates and more.

## Examples

*PUBLISHING_CHANNEL* can be retrieved from your [ContentChef's dashboard](https://app.contentchef.io/).

*PREVIEW_API_KEY* and *ONLINE_API_KEY* must be used to retrieve contents from the `PreviewChannel` and the `OnlineChannel` respectively. Both the api keys can be retrieved from your [ContentChef's dashboard](https://app.contentchef.io/).

Retrieve the *new-header* content from the _live_ environment:

    val onlineContentRequestData = OnlineContentRequestData(  
        "new-header"  
    )
    
    val onlineChannel = contentChef.getOnlineChannel(ONLINE_API_KEY, PUBLISHING_CHANNEL)
    
    onlineChannel.getContent(onlineContentRequestData, {  
      println("onSuccess $it")  
    }, {  
      println("onError $it")  
    })

Preview the *new-header* content in a given future date:

    val targetDate = ContentChefDateFormat.parseDate("2030-01-01T05:42:17.945-05")!!
    
    val previewContentRequestData = PreviewContentRequestData(  
        "new-header", targetDate
    )
    
    val previewChannel = contentChef.getPreviewChannel(PREVIEW_API_KEY, PUBLISHING_CHANNEL)
    
    previewChannel.getContent(previewContentRequestData, {  
      println("onSuccess $it")  
    }, {  
      println("onError $it")  
    })

Search for all the contents with definition as *default-header* in the _live_ environment:

    val searchOnlineRequestData = SearchOnlineRequestData(  
        contentDefinitions = listOf("default-header"),  
      take = 10  
    )
    
    val onlineChannel = contentChef.getOnlineChannel(ONLINE_API_KEY, PUBLISHING_CHANNEL)
    
    onlineChannel.search(searchOnlineRequestData, {  
      println("onSuccess $it")  
    }, {  
      println("onError $it")  
    })

Preview all the contents with definition as *default-header* in a given future date:

    val targetDate = ContentChefDateFormat.parseDate("2030-01-01T05:42:17.945-05")!!
    
    val searchPreviewRequestData = SearchPreviewRequestData(  
        contentDefinitions = listOf("default-header"),  
      targetDate = targetDate,  
      take = 10  
    )
    
    val previewChannel = contentChef.getPreviewChannel(PREVIEW_API_KEY, PUBLISHING_CHANNEL)
    
    previewChannel.search(searchPreviewRequestData, {  
      println("onSuccess $it")  
    }, {  
      println("onError $it")  
    })

Retrieve localized content as *localized-header* from the _live_ environment:

```kotlin
val enOnlineChannel = contentChef.getOnlineChannel(ONLINE_API_KEY, PUBLISHING_CHANNEL, locale = Locale.ENGLISH)
val itOnlineChannel = contentChef.getOnlineChannel(ONLINE_API_KEY, PUBLISHING_CHANNEL, locale = Locale.ITALY)

val localizedHeaderOnlineContentRequestData = OnlineContentRequestData(
    "localized-header"
)

enOnlineChannel.getContent(localizedHeaderOnlineContentRequestData, {
    println("onSuccess $it")
}, { 
    println("onError $it") 
})

itOnlineChannel.getContent(localizedHeaderOnlineContentRequestData, {
    println("onSuccess $it")
}, {
    println("onError $it") 
})
```

Look at *sample-java*, *sample-kt* and *sample-android* in the source code for more examples.

## Installation

To add ContentChef to your project you can use Maven or Gradle.

**Java/Kotlin:**

If using Maven, add this to your `pom.xml` dependencies:

    <dependency>
        <groupId>io.contentchef</groupId>
        <artifactId>contentchef-jvm-callback</artifactId>
        <version>x.y.z</version>
    </dependency>

If using Gradle, add this to your `build.gradle` dependencies:

    implementation 'io.contentchef:contentchef-jvm-callback:x.y.z'

**Android**

Add this to your `build.gradle` dependencies:

    implementation 'io.contentchef:contentchef-jvm-callback-android:x.y.z'

## Dependencies

The Java/Kotlin SDK depends on `org.json:json:20190722`.  
The Android SDK has no dependencies at all.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
## [1.3.0] - 2021-05-14
### Added
- A `Locale` can now be provided to `getPreviewChannel` and `getOnlineChannel` to retrieve localized content
### Fixed
- Fix `NullPointerException` when a 404 response happens. Now `ContentNotFoundException` it's correctly used 

## [1.2.2] - 2021-04-30
### Added
- Now available in MavenCentral
### Fixed
- Fix gradle dependencies sync

## [1.2.1] - 2021-02-04
### Changed
- Using the SDK from Java is now easier thanks to some Java/Kotlin interoperability annotations

## [1.2.0] - 2020-09-04
### Changed
- `previewApiKey` must now be provided to `getPreviewChannel()` instead of `ContentChefEnvironmentConfiguration`
- `onlineApiKey` must now be provided to `getOnlineChannel()` instead of `ContentChefEnvironmentConfiguration`

## [1.1.0] - 2020-07-10
### Changed
- `onlineApiKey` and `previewApiKey` must now be provided to `ContentChefEnvironmentConfiguration`

## [1.0.0] - 2020-01-31
### Added
- Initial version, created by ContentChef team

sample-android/src/main/java/io/contentchef/android/sample/MainActivity.kt
package io.contentchef.android.sample

import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import io.contentchef.callback.android.CallbackContentChefProvider
import io.contentchef.callback.common.ContentChef
import io.contentchef.callback.common.OnlineChannel
import io.contentchef.callback.common.PreviewChannel
import io.contentchef.common.configuration.ContentChefEnvironment
import io.contentchef.common.configuration.ContentChefEnvironmentConfiguration
import io.contentchef.common.request.*
import io.contentchef.common.util.ContentChefDateFormat
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.content_main.*
import java.util.*

class MainActivity : AppCompatActivity() {

    companion object {
        private const val TAG = "MainActivity"
    }

    private lateinit var contentChef: ContentChef
    private lateinit var onlineChannel: OnlineChannel
    private lateinit var previewChannel: PreviewChannel

    //LOCALIZED CHANNELS
    private lateinit var enOnlineChannel: OnlineChannel
    private lateinit var itOnlineChannel: OnlineChannel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)
        contentChef = CallbackContentChefProvider.getContentChef(
            ContentChefEnvironmentConfiguration(ContentChefEnvironment.LIVE, SPACE_ID),
            true
        )

        onlineChannel = contentChef.getOnlineChannel(ONLINE_API_KEY, PUBLISHING_CHANNEL)
        previewChannel = contentChef.getPreviewChannel(PREVIEW_API_KEY, PUBLISHING_CHANNEL)

        fab.setOnClickListener { _ ->

            val previewContentRequestData = PreviewContentRequestData(
                contentIdEditText.text.toString(), Date()
            )

            val onlineContentRequestData = OnlineContentRequestData(
                contentIdEditText.text.toString()
            )

            val targetDate = ContentChefDateFormat.parseDate("2019-11-22T05:42:17.945-05")!!

            val searchPreviewRequestData = SearchPreviewRequestData(
                contentDefinitions = listOf("default-header"),
                targetDate = targetDate
            )

            val searchOnlineRequestData = SearchOnlineRequestData(
                contentDefinitions = listOf("default-header")
            )

            val searchPreviewWithPropFiltersRequestData = SearchPreviewRequestData(
                contentDefinitions = listOf("default-header"),
                propFilters = PropFilters.Builder()
                    .indexedFilterCondition(IndexedFilterCondition.AND)
                    .indexedFilterItem(
                        IndexedFilterItem(
                            IndexedFilterOperator.STARTS_WITH_IC,
                            "header",
                            "A"
                        )
                    )
                    .build()
            )

            previewChannel.getContent(previewContentRequestData, {
                Log.d(TAG, "onSuccess $it")
            }, {
                Log.d(TAG, "onError", it)
            })

            previewChannel.getContent(previewContentRequestData, {
                Log.d(TAG, "onSuccess $it")
            }, {
                Log.d(TAG, "onError", it)
            }, {
                SampleHeader(it.getString("header"))
            })

            onlineChannel.getContent(onlineContentRequestData, {
                Log.d(TAG, "onSuccess $it")
            }, {
                Log.d(TAG, "onError", it)
            })

            onlineChannel.getContent(onlineContentRequestData, {
                Log.d(TAG, "onSuccess $it")
            }, {
                Log.d(TAG, "onError", it)
            }, {
                SampleHeader(it.getString("header"))
            })

            previewChannel.search(searchPreviewRequestData, {
                Log.d(TAG, "onSuccess $it")
            }, {
                Log.d(TAG, "onError", it)
            })

            previewChannel.search(searchPreviewRequestData, {
                Log.d(TAG, "onSuccess $it")
            }, {
                Log.d(TAG, "onError", it)
            }, {
                SampleHeader(it.getString("header"))
            })

            onlineChannel.search(searchOnlineRequestData, {
                Log.d(TAG, "onSuccess $it")
            }, {
                Log.d(TAG, "onError", it)
            })

            onlineChannel.search(searchOnlineRequestData, {
                Log.d(TAG, "onSuccess $it")
            }, {
                Log.d(TAG, "onError", it)
            }, {
                SampleHeader(it.getString("header"))
            })

            //PROP FILTERS EXAMPLE
            previewChannel.search(searchPreviewWithPropFiltersRequestData, {
                Log.d(TAG, "onSuccess $it")
            }, {
                Log.d(TAG, "onError $it")
            })

            previewChannel.search(searchPreviewWithPropFiltersRequestData, {
                Log.d(TAG, "onSuccess $it")
            }, {
                Log.d(TAG, "onError $it")
            }, {
                SampleHeader(it.getString("header"))
            })

            //LOCALIZED CONTENT EXAMPLE
            enOnlineChannel = contentChef.getOnlineChannel(ONLINE_API_KEY, PUBLISHING_CHANNEL, locale = Locale.ENGLISH)
            itOnlineChannel = contentChef.getOnlineChannel(ONLINE_API_KEY, PUBLISHING_CHANNEL, locale = Locale.ITALY)

            val localizedHeaderOnlineContentRequestData = OnlineContentRequestData(
                "test-localized-header"
            )

            enOnlineChannel.getContent(localizedHeaderOnlineContentRequestData, {
                println("onSuccess $it")
            }, {
                println("onError $it")
            })

            itOnlineChannel.getContent(localizedHeaderOnlineContentRequestData, {
                println("onSuccess $it")
            }, {
                println("onError $it")
            })

        }
    }

}
